### PR TITLE
Documenting Optimiser Interface

### PR DESCRIPTION
**Pull request by dhairyagandhi96**
This PR introduces some docs to the optimiser interface, and applies basic fixes to the docstrings for the various optimisers
Ref #871 and #836

### CODE REVIEW

File: src/optimise/optimisers.jl [hunk 1]
```diff
@@ -23,7 +24,8 @@ function apply!(o::Descent, x, Δ)
 end
 
 """
-    Momentum(η = 0.01; ρ = 0.9)
+    Momentum() => Momentum(η, ρ)
```
Comment (mcabbott) [inline src/optimise/optimisers.jl]: Making clear that the parameters aren’t keywords is great, but these `=>` things seem very odd to me. Isn’t it enought to write the defaults next to or below the definition? 
```suggestion
    Momentum(η, ρ) # Defaults: η = 0.01, ρ = 0.9
```

Comment (dhairyagandhi96) [reply src/optimise/optimisers.jl]: I think we also need to make clear that users don't need to specify the parameters if they want to use the defaults. That was kind of the hope with the previous docstring since Julia doesn't do named positional arguments, and defining defaults values like that is pretty common practice.

I'd be happy to make this change if there is consensus.

Comment (mcabbott) [reply src/optimise/optimisers.jl]: Agree with the goals, I just think `=>` is confusing. Maybe another way is to write this: 
```
    Momentum(η, ρ)
    Momentum() # Defaults: η = 0.01, ρ = 0.9
```

Comment (dhairyagandhi96) [reply src/optimise/optimisers.jl]: Right, I think the last commit is close to this, save the `$opt()`

Maybe we add that in

Comment (MikeInnes) [reply src/optimise/optimisers.jl]: Maybe we can just use `=`. That's exactly how you'd define it in Julia anyway, so it should be unambiguous.

Comment (MikeInnes) [reply src/optimise/optimisers.jl]: If we use something like `Defaults`, it should definitely be in the text rather than the signature; perhaps with a description of the parameters. That might actually be better since we can expand eta to "learning rate" etc. Probably it's a good idea to do both things.

File: src/optimise/optimisers.jl [hunk 2]
```diff
@@ -25,7 +27,10 @@ end
 
 """
     Momentum(η, ρ)
-    Defaults: η = 0.01, ρ = 0.9
+
+    Calls to `Momentum()` default to:
```
Comment (MikeInnes) [inline src/optimise/optimisers.jl]: It's still strange to have text / a list inside the code block like this, when it could be normal markdown.

I'm really imagining something like a "parameters" section which outlines the meanings of parameters, and their defaults.

File: src/optimise/optimisers.jl [hunk 3]
```diff
@@ -7,10 +7,29 @@ const ϵ = 1e-8
 # TODO: should use weak refs
 
 """
-    Descent(η)
+  Descent(η)
 
+## Description
```
Comment (MikeInnes) [inline src/optimise/optimisers.jl]: Looks like you have a stray `description` header here.

File: docs/src/training/optimisers.md [hunk 4]
```diff
@@ -58,3 +58,78 @@ AMSGrad
...
+
+```julia
+mutable struct Momentum{T,S,D}
+  eta::T
+  rho::S
+  velocity::D
+end
+```
+
+The `Momentum` type will act as our optimiser in this case. Notice that we have added all the parameters as fields, along with the velocity which we will use as our state. **Note that this behaviour is set to change in consequent versions of Flux**. We can now define the rule applied when this optimiser is invoked.
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]: Reading this, it's not clear what the struct fields actually are. I think it'd be better, for pedagogical purposes, to leave the type parameters off entirely and have an example constructor like `Momentum(eta::Real, rho::Real) = ...`, to show the IdDict being constructed. We can then explain that the IdDict stores a state for each model parameter.

I think the warning is also unnecessary; many things could change, but we'll deprecate those things when we need to.

File: docs/src/training/optimisers.md [hunk 5]
```diff
@@ -58,3 +58,78 @@ AMSGrad
...
+end
+```
+
+This is the basic definition of a Momentum update rule given by:
+$v = ρ * v - η * Δ$
+$w = w - v$
+
+The `apply!` defines the update rules for an optimsier `opt`, given the parameters and gradients. It returns the updated gradients usually. Here, every parameter `x` is retrieved from the running state `v` and subsequently updates the state of the optimiser.
+
+Flux internally calls on this function via the `update!` function. It shares the API with `apply!` but ensures that multiple parameters are handled gracefully. In the future, it will also be delegating immutable update operations.
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]: I would leave the "in future" comment off here too; users here want to understand the current API and I think this might distract from that.

File: docs/src/training/optimisers.md [hunk 6]
```diff
@@ -58,3 +58,78 @@ AMSGrad
...
+
+## Composing Optimisers
+
+Flux defines a special kind of optimiser called simply as `Optimiser` which takes in a arbitrary optimisers as input. Its behaviour is similar to the usual optimisers, but differs in that it acts by calling the optimsers listed in it sequentially. Each optimiser produces a modified gradient
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]:
```suggestion
Flux defines a special kind of optimiser called simply as `Optimiser` which takes in a arbitrary optimisers as input. Its behaviour is similar to the usual optimisers, but differs in that it acts by calling the optimisers listed in it sequentially. Each optimiser produces a modified gradient
```

File: docs/src/training/optimisers.md [hunk 7]
```diff
@@ -58,3 +58,78 @@ AMSGrad
...
+  @. v = ρ * v - η * Δ
+  @. Δ = -v
+end
+```
+
+This is the basic definition of a Momentum update rule given by:
+$v = ρ * v - η * Δ$
+$w = w - v$
+
+The `apply!` defines the update rules for an optimsier `opt`, given the parameters and gradients. It returns the updated gradients usually. Here, every parameter `x` is retrieved from the running state `v` and subsequently updates the state of the optimiser.
```
[the remark this answers is not in the document]
Comment (dhairyagandhi96) [reply docs/src/training/optimisers.md]: Gotcha

File: docs/src/training/optimisers.md [hunk 8]
```diff
@@ -58,3 +58,78 @@ AMSGrad
...
+  η, ρ = o.eta, o.rho
+  v = get!(o.velocity, x, zero(x))::typeof(x)
+  @. v = ρ * v - η * Δ
+  @. Δ = -v
+end
+```
+
+This is the basic definition of a Momentum update rule given by:
+$v = ρ * v - η * Δ$
+$w = w - v$
```
[the remark this answers is not in the document]
Comment (MikeInnes) [reply docs/src/training/optimisers.md]: We probably need to upgrade Documenter to use Katex instead of Mathjax. Can do that as a follow up though

**Comment by dhairyagandhi96** [comment]
bump

File: src/optimise/optimisers.jl [hunk 9]
```diff
@@ -7,10 +7,28 @@ const ϵ = 1e-8
...
+
+opt = Descent(0.3) # use provided η
+
+ps = params(model)
+
+gs = gradient(ps) do
+  loss(x, y)
+end
+
+Flux.Optimise.update(opt, ps, gs)
```
Comment (MikeInnes) [inline src/optimise/optimisers.jl]:
```suggestion
Flux.Optimise.update!(opt, ps, gs)
```

**Comment by dhairyagandhi96** [comment]
Closed via #904
